examples/ASDKgram/Sample/CommentFeedModel.m

```diff
@@ -0,0 +1,196 @@
+{
+  return [_comments objectAtIndex:index];
+}
+
+- (NSUInteger)numberOfCommentsForPhoto
+{
+  return _totalItems;
+}
+
+- (BOOL)numberOfCommentsForPhotoExceedsInteger:(NSUInteger)number
+{
+  return (_totalItems > number);
+}
+
+- (NSAttributedString *)viewAllCommentsAttributedString
+{
+  NSString *string               = [NSString stringWithFormat:@"View all %@ comments", [NSNumber numberWithUnsignedInteger:_totalItems]];
+  NSAttributedString *attrString = [NSAttributedString attributedStringWithString:string fontSize:14 color:[UIColor lightGrayColor] firstWordColor:nil];
+  return attrString;
+}
+
+- (void)requestPageWithCompletionBlock:(void (^)(NSArray *))block
+{
+  // only one fetch at a time
+  if (_fetchPageInProgress) {
+      return;
+  } else {
+    _fetchPageInProgress = YES;
+    [self fetchPageWithCompletionBlock:block];
+  }
+}
+
+- (void)refreshFeedWithCompletionBlock:(void (^)(NSArray *))block
+{
+  // only one fetch at a time
+  if (_refreshFeedInProgress) {
+    return;
+  } else {
+    _refreshFeedInProgress = YES;
+    _currentPage = 0;
+    
+    // FIXME: blow away any other requests in progress
+    
+    [self fetchPageWithCompletionBlock:^(NSArray *newPhotos) {
+      if (block) {
+        block(newPhotos);
+      }
+      _refreshFeedInProgress = NO;
+    } replaceData:YES];
+  }
+}
+
+#pragma mark - Helper Methods
+- (void)fetchPageWithCompletionBlock:(void (^)(NSArray *))block
+{
+  [self fetchPageWithCompletionBlock:block replaceData:NO];
+}
+
+- (void)fetchPageWithCompletionBlock:(void (^)(NSArray *))block replaceData:(BOOL)replaceData
+{
+  // early return if reached end of pages
+  if (_totalPages) {
+    if (_currentPage == _totalPages) {
+      return;
+    }
+  }
+  
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
+    
+    NSMutableArray *newComments = [NSMutableArray array];
+    
+    @synchronized(self) {
+      
+      NSUInteger nextPage = _currentPage + 1;
+      
+      NSString *urlAdditions = [NSString stringWithFormat:@"page=%lu", (unsigned long)nextPage];
+      NSURL *url = [NSURL URLWithString:[_urlString stringByAppendingString:urlAdditions]];
+      
+      NSData *data = [NSData dataWithContentsOfURL:url];
+      
+      if (data) {
+        
+        NSDictionary *response = [NSJSONSerialization JSONObjectWithData:data options:0 error:NULL];
+        
+        if ([response isKindOfClass:[NSDictionary class]]) {
+          
+          _currentPage = [[response valueForKeyPath:@"current_page"] integerValue];
+          _totalPages  = [[response valueForKeyPath:@"total_pages"] integerValue];
+          _totalItems  = [[response valueForKeyPath:@"total_items"] integerValue];
+          
+          NSArray *comments = [response valueForKeyPath:@"comments"];
+          
+          if ([comments isKindOfClass:[NSArray class]]) {
+            
+            NSUInteger numComments = [comments count];
+            if (numComments > 3) {
```

Comment:
@vtsoup it's great to hear that you are investigating framework for your team. I think every team has different requirements, which depend on the design of the app they're creating, the engineers' interest in learning new things, and the size + current level of tech debt in the codebase.

[Sorry this is long and hard to read - I am getting over some RSI right now and using voice dictation, which reads a bit awkwardly...]

The iPhone 6S+ is _the fastest iPhone ever created_, and nearly the fastest iOS device (iPad Pro beats it by less than 30% in CPU!!!)  That said, many apps still have performance problems with it - but because of its sheer power, you need to be looking at a more complex app to see the quite substantial difference.

Pinterest is currently rolling out the first 100%-ASDK version of the app (if you look at it now, you are almost certainly seeing the old version).  The ASDK version is dramatically faster with such a complex app.  Scrolling smoothness is one thing - the old Pinterest does drop frames, sometimes noticably, on the 6S+ although it is not a big problem (the iPhone 6 is much more stuttery, the 6+ is worse, but the 5S and especially 5C and 5 are far worse yet the most common devices...4S is still very common).  Remember iPads are also very resource constrained compared to iPhones, and even the original iPad Air (very very common) is often very with complex apps.

Speaking performance alone - the biggest advantages of ASDK are non-perf -- for Pinterest and many apps, on a fast device like 6S+ the way the app manages loading new content & images is the most influential, esp. on slower connections.

ASDK is not just for accelerating layout and drawing. Although making both of these operations concurrent will also decrease the time to Display content to the user, sometimes meaningfully shortening the transition to a new you or shaving hundreds of milliseconds off of processing a new server response, it is really the interaction-driven network loading that allows the framework to take advantage of whatever quality of network connection it is given.

For example, out-of-the-box using ASNetworkImageNode with nothing more than an alloc/init call, you get these behaviors:

- Automatic preloading, memory caching, and disk caching that is _managed by sensing the user's intent_
- 100% tunable as to when image downloads begin and when their decode starts.
- Powerful placeholder functionality to improve the look and feel of an application even while it is loading data, as well as during decode or drawing operations. (Pinterest uses this for the colored rounded rectangles - nothing more than implementation of one method, -placeholderImage, and even the fading is automatic).
- Multi-directional scrolling support means that Device resources are primarily focused on preparing and storing data related to the direction of user travel. If the user is paging through full-screen views horizontally, the focus will be on loading pages in the direction of travel while maintaining reasonable memory usage of those behind. If the user begins scrolling vertically within a page, memory is freed from some of the surrounding pages, and devices immediately prioritizes loading data in that direction.
- Image downloads are automatically managed at _three_ priority levels, with no implementation work.  Normal preload (low pri), almost onscreen (within decoding range), mid pri, and visible (extreme priority if an image has not finished rendering).
- Native support for Progressive JPEG, which most .jpg files on the internet already support (all good server JPEG encoders create these because they are actually about 10% smaller than non-progressive for same quality).  This intelligently avoids creating progressive scans for images that are off screen, but if the image comes on screen win only part of the data is available, the data downloaded so far it is decoded to create an intermediate quality version and continuously improve as more data arrives. This makes a huge difference on slow connections and for large images.
- Built in GIF support.
- JSON API fetches can also be driven automatically.

This is really just one example of a performance feature of the framework that will affect even the latest devices, and will also remove complexity from your application code. However, it is important to remember that developers are often surrounded by strong Internet connection quality, even the cell networks of the areas we 10 to work. I'm not sure if this is also true in your area, but it is critical to remember that most people using your app have Devices that are at least five times, often 10 times slower in CPU power and a similar amount slower in network performance...

These differences may seem easy to extrapolate, but it is difficult to predict the full impact that a high latency network connection has on the user experience for an application with only basic mechanisms to preload and manage caching. Although it is difficult to test these things yourself, The network connection tool Apple provides can be helpful, and it is definitely recommended that you add performance instrumentation to your app to see real world performance that your users experience.

Despite the performance capabilities of the framework, in the latest versions the best advantage for a team considering adoption include the ability to use a wide library of components with behaviors that significantly add to the Apple APIs. I can't fully elaborate on this feature set right now, but there are many big and small examples.  A small one to check out is .hitTestSlop - trivially make the tap area of _any_ node type bigger in any direction, without overriding something like -pointInside and doing the frame math.

Hope to see you on Slack and answer any further questions.